Adjacent segment compaction

dotnwat opened the pull request:
A simple implementation of compacting adjacent segments. The basic
strategy is to find two segments whose combined size does not exceed
a configurable max threshold size. This is scheduling, and results in a
pair of segments.

This pair of segments is then physically combined to replace the first
segment. The resulting combined segment is then self-compacted as normal
and then redundant second segment is then permenantely removed.

mmaslankaprv commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -9,6 +9,7 @@
 
 #include "storage/disk_log_impl.h"
 
+#include "cluster/namespace.h"
```
I think we do not want `storage` to depend on `cluster` module.

senior7515 replied:
correct!

mmaslankaprv reviewed (commented):
Amazing change, one comment on cluster dependency left.

senior7515 commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -288,11 +289,9 @@ ss::future<> disk_log_impl::gc(compaction_config cfg) {
     if (
-      config().ntp().ns() == redpanda_ignored_ns
-      || config().ntp().ns() == kafka_ignored_ns) {
+      config().ntp().ns() == cluster::redpanda_ns
```
move to the `model::` namespace. doesn't make sense for this cyclical dependency.

senior7515 commented on src/v/config/configuration.cc:
```diff
@@ -396,6 +396,12 @@ configuration::configuration()
       "cache",
       required::no,
       60s)
+  , max_compacted_log_segment_size(
```
Would be good to add this option to the `docs/www/advanced-configuration.md` with a a sentence of context

senior7515 commented on src/v/storage/log_manager.h:
```diff
@@ -96,7 +96,7 @@ struct log_config {
 
     // compacted segment size
-    size_t max_compacted_segment_size = 256_MiB;
+    size_t compacted_segment_size = 256_MiB;
```
beauty!

senior7515 commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -256,10 +259,104 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
+    }
+
+    // the chosen segments all need to be stable
+    const auto unstable = std::any_of(
```
tip-top! love this.

senior7515 commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -256,10 +259,104 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
+    target->mark_as_finished_self_compaction();
+
+    // remove redundant range
+    _segs.erase(std::next(range.first), range.second);
```
does this do the actual optimization of push to either front or back with a `pop_front` or `pop_back` ?

dotnwat replied:
Pretty sure the answer is yes: https://github.com/scylladb/seastar/blob/master/include/seastar/core/circular_buffer.hh#L489

senior7515 commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -256,10 +259,104 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
+    // remove redundant range
+    _segs.erase(std::next(range.first), range.second);
+    for (auto it = std::next(segments.begin()); it != segments.end(); it++) {
+        co_await remove_segment_permanently(*it, "compact_adjacent_segments");
```
interesting. I think should have a comment that while now there is only a for-loop of size 1, this will for for any range.

senior7515 commented on src/v/storage/segment_utils.cc:
```diff
@@ -501,6 +505,45 @@ ss::future<> self_compact_segment(
+    auto output = co_await ss::make_file_output_stream(std::move(writer));
+    for (auto& segment : segments) {
+        auto input = segment->reader().data_stream(0, io_priority);
+        co_await ss::copy(input, output);
```
beauty.

senior7515 commented on src/v/storage/tests/storage_e2e_test.cc:
```diff
@@ -1066,3 +1083,125 @@ FIXTURE_TEST(check_segment_size_jitter, storage_test_fixture) {
       false);
 }
+
+FIXTURE_TEST(adjacent_segment_compaction, storage_test_fixture) {
```
nit: would be good to add a recovery or part of the `op-fuzzer`  operation to ensure that crashing and removing things work.

senior7515 reviewed (approved):
LGTM. 

Very easy to read. Minor cosmetics.

BenPope commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -256,10 +259,104 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
+     * remove index files. they will be rebuilt by the single segment compaction
+     * operation, and also ensures we examine segments correctly on recovery.
+     */
+    if (co_await ss::file_exists(target->index().filename())) {
```
Can compaction fail? Should these be moved to after or just before when mutations to the index occur?

dotnwat replied:
It can fail. Removing these indexes is a way to signal that they should be rebuilt.

BenPope reviewed (commented):
Nice, just a comment about exception safety, which ends up being about not having to rebuild an index if compaction fails.

senior7515 commented:
looks like a real bug

```
ERROR 2021-01-15 04:48:30,186 [shard 0] assert - Assert failure: (/__w/redpanda/redpanda/src/v/storage/opfuzz/opfuzz.cc:267) 'loffsets.dirty_offset == expected' Truncate failed - expected offset 181, have offset {start_offset:0, committed_offset:24, committed_offset_term:0, dirty_offset:24, dirty_offset_term:0, last_term_start_offset:0}
ERROR 2021-01-15 04:48:30,186 [shard 0] assert - Backtrace below:
   0x47e4fb
   0x47e870
   0x47edc1
   0x194a58
   0x1a1367
   0x38e697
   0x38ea91
   0x3bc35d
   0x34cd65
   0x34d94d
   0x1b6666
   0x383401
   /lib/x86_64-linux-gnu/libpthread.so.0+0x958f
   /lib/x86_64-linux-gnu/libc.so.6+0x119222
   --------

```

dotnwat commented:
hmm, weird. there is a race that is known, but opfuzz shouldn't generate it. must be a new thing.

dotnwat commented:
i was able to trigger it. it is the race i had in mind. basically we don't have a good way to protect the segment set itself (e.g. compaction is timer driven and races with things like delete/truncate which manipulate that set). let's chat about that tomorrow. it'd be good to add a fix for that to this pr and then recovery is a separate pr.

senior7515 commented:
> i was able to trigger it. it is the race i had in mind. basically we don't have a good way to protect the segment set itself (e.g. compaction is timer driven and races with things like delete/truncate which manipulate that set). let's chat about that tomorrow. it'd be good to add a fix for that to this pr and then recovery is a separate pr.

makes sense! let's do it.

dotnwat commented:
Changes since last time needed to address failures revealed by opfuzzer:

1. We need to restrict segment pairs to the same term since batches derive their term from their containing file. Not a fundamental limitation, but this is currently an assumption baked in pretty deeply (unless i'm missing something!)

2. I wrongly thought that compaction would operate on the underlying file, but in fact that is naive. Readers limit themselves based on the offset tracker, and so concatenating files is not sufficient. The new approach builds a temporary segment wrapper around the concatenated file and compacts that segment. All the state from this temporary segment is then stolen and used to replace the existing target segment present in the segment set.

Recovery and changes for some of the concurrency scenarios is forthcoming in the next PR.

senior7515 commented:
> Changes since last time needed to address failures revealed by opfuzzer:
> 
> 1. We need to restrict segment pairs to the same term since batches derive their term from their containing file. Not a fundamental limitation, but this is currently an assumption baked in pretty deeply (unless i'm missing something!)
> 2. I wrongly thought that compaction would operate on the underlying file, but in fact that is naive. Readers limit themselves based on the offset tracker, and so concatenating files is not sufficient when compacting the combined segment. The new approach builds a temporary segment wrapper around the concatenated file and compacts that segment. All the state from this temporary segment is then stolen and used to replace the existing target segment present in the segment set.
> 
> Recovery and changes for some of the concurrency scenarios is forthcoming in the next PR.

AHHHH Of course!!!! Good point! hmm... that's problematic. We should allow term compaction as well. But obviously, can be part of a separate PR. I think it's worth opening an issue. On stable clusters, this should be OK for a little while.

senior7515 commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -255,10 +259,145 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
+
+        // found a good range if all the tests pass
+        if (
+          same_term
```
clutch! great catch.

senior7515 reviewed (approved):
Great catch on the term.

senior7515 commented:
Might be worth as a follow up PR w/ the recovery to add an `op` in the `opfuzz` to read+compact in background.

dotnwat commented:
> Might be worth as a follow up PR w/ the recovery to add an `op` in the `opfuzz` to read+compact in background.

For sure. Next PR

dotnwat commented:
> > Changes since last time needed to address failures revealed by opfuzzer:
> > 
> > 1. We need to restrict segment pairs to the same term since batches derive their term from their containing file. Not a fundamental limitation, but this is currently an assumption baked in pretty deeply (unless i'm missing something!)
> > 2. I wrongly thought that compaction would operate on the underlying file, but in fact that is naive. Readers limit themselves based on the offset tracker, and so concatenating files is not sufficient when compacting the combined segment. The new approach builds a temporary segment wrapper around the concatenated file and compacts that segment. All the state from this temporary segment is then stolen and used to replace the existing target segment present in the segment set.
> > 
> > Recovery and changes for some of the concurrency scenarios is forthcoming in the next PR.
> 
> AHHHH Of course!!!! Good point! hmm... that's problematic. We should allow term compaction as well. But obviously, can be part of a separate PR. I think it's worth opening an issue. On stable clusters, this should be OK for a little while.

@mmaslankaprv pointed out that the term-per-file was motivated for a desire of having fast truncations. seems we could also do this with some indexing if we in fact need more flexibility to store batches from multiple terms in a single file.
